brainiak/matnorm_models/base.py

```diff
@@ -0,0 +1,229 @@
+import tensorflow as tf
+from sklearn.base import BaseEstimator
+
+
+class MatnormModelBase(BaseEstimator):
+    """ Base class implementing shared functionality for matrix-variate normal
+     models in tensorflow.
+
+    Not intended to be used directly -- contains a wrapper for the TF optimizers
+    that does convergence checks, plus likelihood for mnorm (incl. with marginal
+    and conditional densities).
+    """
+
+    def _optimize_impl(self, optfun, loss, optvars, feed_dict, max_iter, step,
+                       loss_tol, grad_tol):
+        """Implementation method for optimization that does
+            convergence checks and output
+        """
+
+        grad = tf.concat(0, [tf.reshape(g[0], [-1]) for g in
+                         self.optimizer.compute_gradients(loss, optvars)])
+
+        max_abs_current_grad_op = tf.reduce_max(tf.abs(grad))
+
+        past_loss = 0
+        current_loss = self.sess.run(loss, feed_dict=feed_dict)
+
+        for n in range(max_iter):
+            self.sess.run(optfun, feed_dict=feed_dict)
+            past_loss = current_loss
+            current_loss = self.sess.run(loss, feed_dict=feed_dict)
+            max_abs_current_grad = self.sess.run(max_abs_current_grad_op,
+                                                 feed_dict=feed_dict)
+
+            # check tolerances
+            if abs((current_loss - past_loss) /
+                    max(current_loss, past_loss)) < loss_tol:
+                print('loss tolerance reached on iter %i, %f, stopping' %
```

Comment:
Indeed, we do not use `print` in BrainIAK. This used to be checked automatically, but the tool broke:
https://github.com/JBKahn/flake8-print/pull/17